Refactor: Improve API design of `ScriptToUniv`, `TxToUniv` etc to return the `UniValue` instead of mutating a parameter/reference

mjdietzx opened the pull request:
Now functions return a `UniValue` instead of mutating a parameter that's passed by reference.

After this PR, any time a `UniValue` is passed by reference, it should be `const`. This is the case everywhere in the codebase now (please double check me on this) except for `SoftForkDescPushBack` (which I didn't want to touch), and the RPC request handling logic.

Motivation for this PR was based on a suggestion in #22924 review:
> I wonder why we don't return the UniValue (or optional<UniValue> in case it can fail) instead of having it as second parameter. It would seem better API design to me.

I figured it's simple enough, why not do this across the entire codebase. Hopefully I didn't get carried away

mjdietzx committed:
refactor: `ScriptToUniv` return `UniValue` instead of mutating reference

mjdietzx committed:
refactor: `TxToUniv` return `UniValue` instead of mutating reference

lsilva01 commented:
Concept ACK.

mjdietzx committed:
refactor: `TxToJSON` return `UniValue` instead of mutating reference

mjdietzx committed:
refactor: prefer snake case, hashBlock renamed block_hash in rpc/rawtransaction.cpp

mjdietzx committed:
refactor: remove code duplication b/w `TxToJSON` and `TxToUniv`

mjdietzx committed:
refactor: `entryToJSON` return `UniValue` instead of mutating reference

mjdietzx committed:
refactor: `SignTransaction` return `UniValue` instead of mutating reference

mjdietzx committed:
refactor: `SignTransactionResultToJSON` return `UniValue` instead of mutating reference

mjdietzx committed:
refactor: `TxInErrorToJSON` return `UniValue` instead of mutating reference

mjdietzx committed:
refactor: `WalletTxToJSON` return `UniValue` instead of mutating reference

mjdietzx committed:
refactor: `ProcessSubScript` return `UniValue` instead of mutating reference

mjdietzx committed:
refactor: `GetWalletBalances` return `UniValue` instead of mutating reference

mjdietzx committed:
refactor: `ParseGetInfoResult` return a value instead of mutating `UniValue` reference

mjdietzx committed:
refactor: `RPCExecutor::request` catch by reference can be const

MarcoFalke commented on src/test/fuzz/script.cpp:
```diff
@@ -108,12 +108,7 @@ FUZZ_TARGET_INIT(script, initialize_script)
     (void)ScriptToAsmStr(script, false);
     (void)ScriptToAsmStr(script, true);
```
unrelated, but could also use ConsumeBool here

mjdietzx replied:
This has been done in this commit fae3f178238df96554dc2495e040f5580b55408a

Marking as resolved.

rajarshimaitra reviewed (approved):
Concept + tACK https://github.com/bitcoin/bitcoin/pull/23507/commits/092ad6a03090172e61ac95a328b9d429e043e8ff

Agree that this makes code much simpler. Never was a fan of argument modification. 😅

Tested that all tests are passing. No unexpected behavior change.

fanquake commented:
I'm going to close this for now. Let's re-open / followup once we figure out the status of, and potentially merge #22924. (I'm doing a rebase at the moment).

mjdietzx commented:
Sounds good, I'll start my review of #24673 today, and will rebase this on top of that PR if it makes sense - then we can see where this PR goes from there

fanquake commented:
#24673 has now been merged.

mjdietzx commented:
I've rebased this PR, but I have not force pushed yet. IIRC I need to make sure I re-open the PR before force pushing. But, I can't figure out how to re-open this PR (I think it may be due to "This pull request is closed, but the mjdietzx:ret_UniValue_instead_of_pass_by_ref branch has unmerged commits.").

Can anyone advise on how I should move forward? Should I create a new branch and just reference this PR (leave it closed)?

fanquake commented:
@mjdietzx you should be able to force push to the branch now.

laanwj commented:
Concept ACK. Returning out-only return values n is a clear improvement in API usability.

mjdietzx commented:
Does anyone want this PR enough to review it if I rebase? If so I will do a careful rebase/review. If not I'm ok with closing this

fanquake commented:
> Does anyone want this PR enough to review it if I rebase? If so I will do a careful rebase/review. If not I'm ok with closing this

I think this change is worthwhile. If you're happy to do a rebase I'll review it. If not, I'll pick this up.

mjdietzx commented:
Thanks @fanquake, rebased.

I am also going to do a thorough re-review as well since I did this a while ago. Also I dropped fcffbdefd2d5f5ec9bba59f55f5b1db6e9a9965c for now, I need to take some time to go through @MarcoFalke's https://github.com/bitcoin/bitcoin/pull/25464 before I rebase this commit.
